File: app/src/main/java/com/blues/module_study/model_fuli/view/FuliFragment.java
package com.blues.module_study.model_fuli.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.LinearLayout;

import com.blues.framework.base.BaseFragment;
import com.blues.R;
import com.blues.module_study.model_fuli.FuliContract;
import com.blues.module_study.model_fuli.model.FuliEntity;
import com.blues.module_study.model_fuli.presenter.FuliPresenter;
import com.jude.easyrecyclerview.EasyRecyclerView;
import com.jude.easyrecyclerview.adapter.BaseViewHolder;
import com.jude.easyrecyclerview.adapter.RecyclerArrayAdapter;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.header.ClassicsHeader;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;

import java.util.List;

public class FuliFragment extends BaseFragment implements FuliContract.iContractView, OnRefreshListener {

    private SmartRefreshLayout fuliSr;
    private ViewStub networkVS;
    private View v;
    private FuliContract.iContractPresenter iPresenter;
    private RecyclerArrayAdapter<FuliEntity.ResultsBean> mAdapter;

    @Override
    public int setLayoutResourceId() {
        return R.layout.fragment_fuli;
    }

    @Override
    public void initLayout(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        EasyRecyclerView fuliErv = rootView.findViewById(R.id.rv_fuli);
        fuliSr = rootView.findViewById(R.id.sr_fuli);
        networkVS = rootView.findViewById(R.id.view_stub_network);
        //设置 Header 为 经典 样式 带最后刷新时间
        fuliSr.setRefreshHeader(new ClassicsHeader(getmContext()).setEnableLastTime(true));
        fuliSr.setEnableHeaderTranslationContent(true);

        mPresenter = new FuliPresenter(this);

        iPresenter.initData();

        fuliErv.setLayoutManager(new LinearLayoutManager(getmContext()));
        fuliErv.setAdapter(mAdapter = new RecyclerArrayAdapter<FuliEntity.ResultsBean>(getmContext()) {
            @Override
            public BaseViewHolder OnCreateViewHolder(ViewGroup parent, int viewType) {
                return new FuliViewHolder(parent);
            }
        });
        mAdapter.setMore(R.layout.view_loading_more, new RecyclerArrayAdapter.OnMoreListener() {
            @Override
            public void onMoreShow() {
                iPresenter.loadMore();
            }

            @Override
            public void onMoreClick() {

            }
        });
        mAdapter.setNoMore(R.layout.view_load_no_more);

    }

    @Override
    public void lazyFetchData() {

    }

    @Override
    public void setListener() {
        fuliSr.setEnableRefresh(true);
        fuliSr.setOnRefreshListener(this);
    }

    @Override
    public void setData(List<FuliEntity.ResultsBean> result) {

        if (v != null) {
            v.setVisibility(View.GONE);
        }

        if (result != null) {
            for (FuliEntity.ResultsBean ios : result) {

                mAdapter.add(ios);
            }

        }
        mAdapter.notifyDataSetChanged();

    }

    @Override
    public void setPresenter(FuliContract.iContractPresenter presenter) {
        this.iPresenter = presenter;
    }

    @Override
    public void showBegin() {

    }

    @Override
    public void showLoading() {

    }

    @Override
    public void showFinished() {
        fuliSr.finishRefresh();
    }

    @Override
    public void showError() {
        try {
            v = networkVS.inflate();
            LinearLayout blankLl = v.findViewById(R.id.error_blank);
            blankLl.setOnClickListener(v1 -> {
                mAdapter.clear();
                iPresenter.initData();
            });
        } catch (Exception e) {
            e.printStackTrace();
            v.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void showEmpty() {

    }

    @Override
    public void onRefresh(@NonNull RefreshLayout refreshLayout) {
        mAdapter.clear();
        iPresenter.initData();
    }
}

File: app/src/main/java/com/blues/gankio/v1/view/FuliFragment.kt
package com.blues.gankio.v1.view

import com.scwang.smartrefresh.layout.SmartRefreshLayout
import android.view.ViewStub
import com.jude.easyrecyclerview.adapter.RecyclerArrayAdapter
import com.blues.R
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import androidx.lifecycle.lifecycleScope
import com.jude.easyrecyclerview.EasyRecyclerView
import com.scwang.smartrefresh.layout.header.ClassicsHeader
import androidx.recyclerview.widget.LinearLayoutManager
import com.jude.easyrecyclerview.adapter.RecyclerArrayAdapter.OnMoreListener
import com.blues.framework.base.BaseKoinFragment
import com.blues.gankio.v1.api.GankType
import com.blues.gankio.v1.model.GankBean
import com.blues.gankio.v1.vm.GankViewModel
import com.jude.easyrecyclerview.adapter.BaseViewHolder
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.scwang.smartrefresh.layout.listener.OnRefreshListener
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel

class FuliFragment : BaseKoinFragment(), OnRefreshListener {

    private val fuliViewModel: GankViewModel by viewModel()

    private lateinit var fuliSr: SmartRefreshLayout
    private var networkVS: ViewStub? = null
    private lateinit var mAdapter: RecyclerArrayAdapter<GankBean.ResultsBean>

    override fun collect() {
        lifecycleScope.launch {
            fuliViewModel.listData.collect {
                it.results.let { list ->
                    mAdapter.apply {
                        addAll(list)
                        notifyDataSetChanged()
                    }
                    fuliSr.finishRefresh()
                }
            }
        }
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        mAdapter.clear()
        fuliViewModel.initData(GankType.PICTURE.type)
    }

    override fun getLayoutId(): Int = R.layout.fragment_fuli

    override fun initData(inflater: LayoutInflater, container: ViewGroup?, saveInstanced: Bundle?) {

        val fuliErv: EasyRecyclerView = rootView.findViewById(R.id.rv_fuli)
        fuliSr = rootView.findViewById<SmartRefreshLayout>(R.id.sr_fuli).apply {

            //设置 Header 为 经典 样式 带最后刷新时间
            setRefreshHeader(ClassicsHeader(requireContext()).setEnableLastTime(true))
            setEnableHeaderTranslationContent(true)
            setEnableRefresh(true)
            setOnRefreshListener(this@FuliFragment)
        }
        onRefresh(fuliSr)

        networkVS = rootView.findViewById(R.id.view_stub_network)

        fuliErv.setLayoutManager(LinearLayoutManager(requireContext()))
        fuliErv.adapter = object : RecyclerArrayAdapter<GankBean.ResultsBean>(requireContext()) {
            override fun OnCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<*> {
                return FuliViewHolder(parent)
            }
        }.also { mAdapter = it }

        mAdapter.apply {
            setNoMore(R.layout.view_load_no_more)
            setMore(R.layout.view_loading_more, object : OnMoreListener {
                override fun onMoreShow() {
                    fuliViewModel.loadMore(GankType.PICTURE.type)
                }

                override fun onMoreClick() {}
            })
        }
    }
}